### ISSUE TITLE
Unable to make socket work

### ISSUE THREAD

**Issue by axeloz** [REBELinBLUE/deployer#388]
Before submitting your issue, please make sure that you've checked all of the checkboxes below.

- [x] You're running the [latest release](https://github.com/REBElinBLUE/deployer/releases/latest) version of Deployer.
- [x] Ensure that you're running at least PHP 7.0.8, you can check this by running `php -v`
- [x] You've ran `composer install --no-dev` from the root of your installation.
- [x] You've ran `npm install --production` from the root of your installation.
- [x] You've ran `rm -rf bootstrap/cache/*` from the root of your installation.
- [x] You have restarted the queue listener and node socket server.

### Describe the bug

Sorry for asking, I should find the solution but I've been looking around for some time now and you may have a quickfix in mind. 

The socket is not working, it get the red error banner message. It used to work just fine.
`A connection could not be established to the socket at https://deploy.xxx.xxxxx.media` 
I'm using Supervisor

In Nginx logs, I have:
```
2018/07/31 17:09:53 [error] 19200#19200: *15072 connect() failed (111: Connection refused) while connecting to upstream, client: 85.171.47.225, server: deploy.xxx.media, request: "GET /socket.
io/?jwt=eyJ0eXAiOiJKV1QiTUsImp0aSI6Ik16SkViR1ZEWnpRMVFuRTRTMDlTTWtWMWVrUmpVM1J5WVVaMmJEI6ycIIEbGXoE-FU_DHu1Uk&EIO=3&transport=websocket HTTP/1.1", upstream: "http://127.0.0.1:6001/socket.
io/?jwt=eyJ0iZiI6MTUzMzA0OTc1NSwiZXhwIjoxNTMzMDYwNTU1LCJkYXRhIjp7InVzZXJJZCI6MX0sInN1YiI6MX0.xq066QLVUQe62uZBlBc87EdycIIEbGXoE-FU_DHu1Uk&EIO=3&transport=websocket"
```

The upstream is `http://127.0.0.1:6001/socket`
When logged in on the server, a `telnet localhost 6001` seems to work with no output though.
```
root@dev:/tank/www/deployer# telnet localhost 6001
Trying ::1...
Connected to localhost.
```
Supervisor seems to work too:
```
root@dev:~# supervisorctl
deployer:queue_0                 RUNNING   pid 20212, uptime 0:01:23
deployer:queue_1                 RUNNING   pid 20211, uptime 0:01:23
deployer:queue_2                 RUNNING   pid 20210, uptime 0:01:23
deployer-broadcast               RUNNING   pid 20213, uptime 0:01:23
deployer-socketio                RUNNING   pid 20214, uptime 0:01:23
```

Thanks

**Comment by REBELinBLUE** [comment]
I think you need to change your nginx config to use the IP, just 127.0.0.1 for the node server, not the hostname

**Comment by REBELinBLUE** [comment]
oops, sorry just realised `server` is just the server name, the upstream is the IP.


Hmm

**Comment by REBELinBLUE** [comment]
Can you try starting the node server with `DEBUG=*` https://socket.io/docs/logging-and-debugging/ and see if there is any additional output in the supervisor log

**Comment by axeloz** [comment]
I forgot to mention that deployments ARE working though. I just need to reload the page to check the status.

**Comment by REBELinBLUE** [comment]
Hmm, this is really odd. Is there anything in the Redis log?

Yeah, deployments will continue to work, the websocket is only used for broadcasting updates to the client instead of doing AJAX polling.

**Comment by axeloz** [comment]
OK I found the problem: 

my instance of deployer runs in HTTPS with Let's Encrypt on top of Nginx. Because of this, my socket URL into the .env file is https:// too. 
But in socket.js, I found these lines:
```
if (/^https/i.test(process.env.SOCKET_URL)) {
  const sslConfig = { ... }
}
```

So socket.io tries to create an SSL connection when Nginx is already handling this as a proxy. 
The connection between Nginx and Socket.io is not supposed to use SSL, at least in my case because they use localhost, on the same server and because the 6001 port is closed in my firewall from the outside. 

To fix it permanently, I guess there should be a new variable into the .env file:
`SOCKET_SSL=true|false`
and replace the previous source lines by:
```
if (process.env.SOCKET_SSL == true) {
  const sslConfig = { ... }
}
```
That would allow users to choose whether Nginx or Socket.io should handle the SSL connection or not.
That is what I did on my side and it works just fine.

Thanks
